HCA Harmonised
================

Load the package

``` r
library(HCAquery)
library(dplyr)
library(dbplyr)
library(SingleCellExperiment)
```

Load the metadata

``` r
get_metadata()
#> # Source:   table<metadata> [?? x 56]
#> # Database: sqlite 3.40.0 [/vast/scratch/users/milton.m/cache/hca_harmonised/metadata.sqlite]
#>    .cell   sampl‚Ä¶¬π .sample .samp‚Ä¶¬≤ assay assay‚Ä¶¬≥ file_‚Ä¶‚Å¥ cell_‚Ä¶‚Åµ cell_‚Ä¶‚Å∂ devel‚Ä¶‚Å∑ devel‚Ä¶‚Å∏ disease disea‚Ä¶‚Åπ ethni‚Ä¶Àü ethni‚Ä¶Àü file_id is_pr‚Ä¶Àü organ‚Ä¶Àü organ‚Ä¶Àü sampl‚Ä¶Àü sex   sex_o‚Ä¶Àü tissue
#>    <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr> 
#>  1 AAACCT‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  2 AAACCT‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  3 AAACCT‚Ä¶ 02eb2e‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 30f754‚Ä¶ lumina‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  4 AAACCT‚Ä¶ 02eb2e‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 30f754‚Ä¶ lumina‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  5 AAACCT‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  6 AAACCT‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  7 AAACCT‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  8 AAACGG‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#>  9 AAACGG‚Ä¶ 02eb2e‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 30f754‚Ä¶ lumina‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#> 10 AAACGG‚Ä¶ 8a0fe0‚Ä¶ 5f20d7‚Ä¶ D17PrP‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ 1e334b‚Ä¶ basal ‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ Europe‚Ä¶ HANCES‚Ä¶ 00d626‚Ä¶ FALSE   Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ perip‚Ä¶
#> # ‚Ä¶ with more rows, 33 more variables: tissue_ontology_term_id <chr>, tissue_harmonised <chr>, age_days <dbl>, dataset_id <chr>, collection_id <chr>, cell_count <int>,
#> #   dataset_deployments <chr>, is_primary_data.y <chr>, is_valid <int>, linked_genesets <int>, mean_genes_per_cell <dbl>, name <chr>, published <int>, revision <int>,
#> #   schema_version <chr>, tombstone <int>, x_normalization <chr>, created_at.x <dbl>, published_at <dbl>, revised_at <dbl>, updated_at.x <dbl>, filename <chr>, filetype <chr>,
#> #   s3_uri <chr>, user_submitted <int>, created_at.y <dbl>, updated_at.y <dbl>, cell_type_harmonised <chr>, confidence_class <dbl>, cell_annotation_azimuth_l2 <chr>,
#> #   cell_annotation_blueprint_singler <chr>, n_cell_type_in_tissue <int>, n_tissue_in_cell_type <int>, and abbreviated variable names ¬π‚Äãsample_id_db, ¬≤‚Äã.sample_name,
#> #   ¬≥‚Äãassay_ontology_term_id, ‚Å¥‚Äãfile_id_db, ‚Åµ‚Äãcell_type, ‚Å∂‚Äãcell_type_ontology_term_id, ‚Å∑‚Äãdevelopment_stage, ‚Å∏‚Äãdevelopment_stage_ontology_term_id, ‚Åπ‚Äãdisease_ontology_term_id, Àü‚Äãethnicity,
#> #   Àü‚Äãethnicity_ontology_term_id, Àü‚Äãis_primary_data.x, Àü‚Äãorganism, Àü‚Äãorganism_ontology_term_id, Àü‚Äãsample_placeholder, Àü‚Äãsex_ontology_term_id
```

Explore the HCA content

``` r
get_metadata() |>
    distinct(tissue, file_id) |>
    count(tissue) |>
    arrange(desc(n))
#> # Source:     SQL [?? x 2]
#> # Database:   sqlite 3.40.0 [/vast/scratch/users/milton.m/cache/hca_harmonised/metadata.sqlite]
#> # Ordered by: desc(n)
#>    tissue                    n
#>    <chr>                 <int>
#>  1 blood                    47
#>  2 heart left ventricle     46
#>  3 cortex of kidney         31
#>  4 renal medulla            29
#>  5 lung                     27
#>  6 middle temporal gyrus    24
#>  7 liver                    24
#>  8 kidney                   19
#>  9 intestine                18
#> 10 thymus                   17
#> # ‚Ä¶ with more rows
```

Query raw counts

``` r
sce <-
    get_metadata() |>
    filter(
        ethnicity == "African" &
            assay %LIKE% "%10x%" &
            tissue == "lung parenchyma" &
            cell_type %LIKE% "%CD4%"
    ) |>
    get_SingleCellExperiment()
#> ‚Ñπ Realising metadata.
#> ‚Ñπ Synchronising files
#> ‚Ñπ Compiling Single Cell Experiment.
#> ‚Ñπ Attaching metadata.

sce
#> # A SingleCellExperiment-tibble abstraction: 1,571 √ó 56
#> # [90mFeatures=60661 | Cells=1571 | Assays=counts, cpm[0m
#>    .cell   sampl‚Ä¶¬π .sample .samp‚Ä¶¬≤ assay assay‚Ä¶¬≥ file_‚Ä¶‚Å¥ cell_‚Ä¶‚Åµ cell_‚Ä¶‚Å∂ devel‚Ä¶‚Å∑ devel‚Ä¶‚Å∏ disease disea‚Ä¶‚Åπ ethni‚Ä¶Àü ethni‚Ä¶Àü file_id is_pr‚Ä¶Àü organ‚Ä¶Àü organ‚Ä¶Àü sampl‚Ä¶Àü sex   sex_o‚Ä¶Àü tissue
#>    <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr> 
#>  1 ACAGCC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  2 GGGAAT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  3 TCTTCG‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  4 CCTTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  5 ATCTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  6 CATCAG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  7 AGTCTT‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  8 CGCGTT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  9 TGGCTG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#> 10 TTCCCA‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#> # ‚Ä¶ with 1,561 more rows, 33 more variables: tissue_ontology_term_id <chr>, tissue_harmonised <chr>, age_days <dbl>, dataset_id <chr>, collection_id <chr>, cell_count <int>,
#> #   dataset_deployments <chr>, is_primary_data.y <chr>, is_valid <int>, linked_genesets <int>, mean_genes_per_cell <dbl>, name <chr>, published <int>, revision <int>,
#> #   schema_version <chr>, tombstone <int>, x_normalization <chr>, created_at.x <dbl>, published_at <dbl>, revised_at <dbl>, updated_at.x <dbl>, filename <chr>, filetype <chr>,
#> #   s3_uri <chr>, user_submitted <int>, created_at.y <dbl>, updated_at.y <dbl>, cell_type_harmonised <chr>, confidence_class <dbl>, cell_annotation_azimuth_l2 <chr>,
#> #   cell_annotation_blueprint_singler <chr>, n_cell_type_in_tissue <int>, n_tissue_in_cell_type <int>, and abbreviated variable names ¬π‚Äãsample_id_db, ¬≤‚Äã.sample_name,
#> #   ¬≥‚Äãassay_ontology_term_id, ‚Å¥‚Äãfile_id_db, ‚Åµ‚Äãcell_type, ‚Å∂‚Äãcell_type_ontology_term_id, ‚Å∑‚Äãdevelopment_stage, ‚Å∏‚Äãdevelopment_stage_ontology_term_id, ‚Åπ‚Äãdisease_ontology_term_id, Àü‚Äãethnicity,
#> #   Àü‚Äãethnicity_ontology_term_id, Àü‚Äãis_primary_data.x, Àü‚Äãorganism, Àü‚Äãorganism_ontology_term_id, Àü‚Äãsample_placeholder, Àü‚Äãsex_ontology_term_id
```

Query counts scaled per million. This is helpful if just few genes are
of interest

``` r
sce <-
    get_metadata() |>
    filter(
        ethnicity == "African" &
            assay %LIKE% "%10x%" &
            tissue == "lung parenchyma" &
            cell_type %LIKE% "%CD4%"
    ) |>
    get_SingleCellExperiment(assays = "cpm")
#> ‚Ñπ Realising metadata.
#> ‚Ñπ Synchronising files
#> ‚Ñπ Compiling Single Cell Experiment.
#> ‚Ñπ Attaching metadata.

sce
#> # A SingleCellExperiment-tibble abstraction: 1,571 √ó 56
#> # [90mFeatures=60661 | Cells=1571 | Assays=cpm[0m
#>    .cell   sampl‚Ä¶¬π .sample .samp‚Ä¶¬≤ assay assay‚Ä¶¬≥ file_‚Ä¶‚Å¥ cell_‚Ä¶‚Åµ cell_‚Ä¶‚Å∂ devel‚Ä¶‚Å∑ devel‚Ä¶‚Å∏ disease disea‚Ä¶‚Åπ ethni‚Ä¶Àü ethni‚Ä¶Àü file_id is_pr‚Ä¶Àü organ‚Ä¶Àü organ‚Ä¶Àü sampl‚Ä¶Àü sex   sex_o‚Ä¶Àü tissue
#>    <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr> 
#>  1 ACAGCC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  2 GGGAAT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  3 TCTTCG‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  4 CCTTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  5 ATCTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  6 CATCAG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  7 AGTCTT‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  8 CGCGTT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  9 TGGCTG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#> 10 TTCCCA‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#> # ‚Ä¶ with 1,561 more rows, 33 more variables: tissue_ontology_term_id <chr>, tissue_harmonised <chr>, age_days <dbl>, dataset_id <chr>, collection_id <chr>, cell_count <int>,
#> #   dataset_deployments <chr>, is_primary_data.y <chr>, is_valid <int>, linked_genesets <int>, mean_genes_per_cell <dbl>, name <chr>, published <int>, revision <int>,
#> #   schema_version <chr>, tombstone <int>, x_normalization <chr>, created_at.x <dbl>, published_at <dbl>, revised_at <dbl>, updated_at.x <dbl>, filename <chr>, filetype <chr>,
#> #   s3_uri <chr>, user_submitted <int>, created_at.y <dbl>, updated_at.y <dbl>, cell_type_harmonised <chr>, confidence_class <dbl>, cell_annotation_azimuth_l2 <chr>,
#> #   cell_annotation_blueprint_singler <chr>, n_cell_type_in_tissue <int>, n_tissue_in_cell_type <int>, and abbreviated variable names ¬π‚Äãsample_id_db, ¬≤‚Äã.sample_name,
#> #   ¬≥‚Äãassay_ontology_term_id, ‚Å¥‚Äãfile_id_db, ‚Åµ‚Äãcell_type, ‚Å∂‚Äãcell_type_ontology_term_id, ‚Å∑‚Äãdevelopment_stage, ‚Å∏‚Äãdevelopment_stage_ontology_term_id, ‚Åπ‚Äãdisease_ontology_term_id, Àü‚Äãethnicity,
#> #   Àü‚Äãethnicity_ontology_term_id, Àü‚Äãis_primary_data.x, Àü‚Äãorganism, Àü‚Äãorganism_ontology_term_id, Àü‚Äãsample_placeholder, Àü‚Äãsex_ontology_term_id
```

Extract only a subset of genes:

``` r
get_metadata() |>
    filter(
        ethnicity == "African" &
            assay %LIKE% "%10x%" &
            tissue == "lung parenchyma" &
            cell_type %LIKE% "%CD4%"
    ) |>
    get_SingleCellExperiment(features = "PUM1")
#> ‚Ñπ Realising metadata.
#> ‚Ñπ Synchronising files
#> ‚Ñπ Compiling Single Cell Experiment.
#> ‚Ñπ Attaching metadata.
#> # A SingleCellExperiment-tibble abstraction: 1,571 √ó 56
#> # [90mFeatures=1 | Cells=1571 | Assays=counts, cpm[0m
#>    .cell   sampl‚Ä¶¬π .sample .samp‚Ä¶¬≤ assay assay‚Ä¶¬≥ file_‚Ä¶‚Å¥ cell_‚Ä¶‚Åµ cell_‚Ä¶‚Å∂ devel‚Ä¶‚Å∑ devel‚Ä¶‚Å∏ disease disea‚Ä¶‚Åπ ethni‚Ä¶Àü ethni‚Ä¶Àü file_id is_pr‚Ä¶Àü organ‚Ä¶Àü organ‚Ä¶Àü sampl‚Ä¶Àü sex   sex_o‚Ä¶Àü tissue
#>    <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr> <chr>   <chr> 
#>  1 ACAGCC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  2 GGGAAT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  3 TCTTCG‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  4 CCTTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  5 ATCTAC‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  6 CATCAG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  7 AGTCTT‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#>  8 CGCGTT‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#>  9 TGGCTG‚Ä¶ c7f14e‚Ä¶ 9ef5ea‚Ä¶ VUHD69‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 31-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    fema‚Ä¶ PATO:0‚Ä¶ lung ‚Ä¶
#> 10 TTCCCA‚Ä¶ 33cdeb‚Ä¶ 4fc10a‚Ä¶ VUHD92‚Ä¶ 10x ‚Ä¶ EFO:00‚Ä¶ bc380d‚Ä¶ CD4-po‚Ä¶ CL:000‚Ä¶ 55-yea‚Ä¶ HsapDv‚Ä¶ normal  PATO:0‚Ä¶ African HANCES‚Ä¶ 6661ab‚Ä¶ TRUE    Homo s‚Ä¶ NCBITa‚Ä¶ <NA>    male  PATO:0‚Ä¶ lung ‚Ä¶
#> # ‚Ä¶ with 1,561 more rows, 33 more variables: tissue_ontology_term_id <chr>, tissue_harmonised <chr>, age_days <dbl>, dataset_id <chr>, collection_id <chr>, cell_count <int>,
#> #   dataset_deployments <chr>, is_primary_data.y <chr>, is_valid <int>, linked_genesets <int>, mean_genes_per_cell <dbl>, name <chr>, published <int>, revision <int>,
#> #   schema_version <chr>, tombstone <int>, x_normalization <chr>, created_at.x <dbl>, published_at <dbl>, revised_at <dbl>, updated_at.x <dbl>, filename <chr>, filetype <chr>,
#> #   s3_uri <chr>, user_submitted <int>, created_at.y <dbl>, updated_at.y <dbl>, cell_type_harmonised <chr>, confidence_class <dbl>, cell_annotation_azimuth_l2 <chr>,
#> #   cell_annotation_blueprint_singler <chr>, n_cell_type_in_tissue <int>, n_tissue_in_cell_type <int>, and abbreviated variable names ¬π‚Äãsample_id_db, ¬≤‚Äã.sample_name,
#> #   ¬≥‚Äãassay_ontology_term_id, ‚Å¥‚Äãfile_id_db, ‚Åµ‚Äãcell_type, ‚Å∂‚Äãcell_type_ontology_term_id, ‚Å∑‚Äãdevelopment_stage, ‚Å∏‚Äãdevelopment_stage_ontology_term_id, ‚Åπ‚Äãdisease_ontology_term_id, Àü‚Äãethnicity,
#> #   Àü‚Äãethnicity_ontology_term_id, Àü‚Äãis_primary_data.x, Àü‚Äãorganism, Àü‚Äãorganism_ontology_term_id, Àü‚Äãsample_placeholder, Àü‚Äãsex_ontology_term_id
```

Extract the counts as a Seurat object:

``` r
get_metadata() |>
    filter(
        ethnicity == "African" &
            assay %LIKE% "%10x%" &
            tissue == "lung parenchyma" &
            cell_type %LIKE% "%CD4%"
    ) |>
    get_seurat()
#> ‚Ñπ Realising metadata.
#> ‚Ñπ Synchronising files
#> ‚Ñπ Compiling Single Cell Experiment.
#> ‚Ñπ Attaching metadata.
#> Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
#> An object of class Seurat 
#> 60661 features across 1571 samples within 1 assay 
#> Active assay: originalexp (60661 features, 0 variable features)
```

``` r
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: CentOS Linux 7 (Core)
#> 
#> Matrix products: default
#> BLAS:   /stornext/System/data/apps/R/R-4.2.1/lib64/R/lib/libRblas.so
#> LAPACK: /stornext/System/data/apps/R/R-4.2.1/lib64/R/lib/libRlapack.so
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#>  [1] SingleCellExperiment_1.18.1 dbplyr_2.2.1                dplyr_1.0.10                HCAquery_0.1.0              SummarizedExperiment_1.26.1 Biobase_2.56.0             
#>  [7] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4         IRanges_2.30.1              S4Vectors_0.34.0            BiocGenerics_0.42.0         MatrixGenerics_1.8.1       
#> [13] matrixStats_0.63.0          testthat_3.1.6             
#> 
#> loaded via a namespace (and not attached):
#>   [1] utf8_1.2.2                     spatstat.explore_3.0-5         reticulate_1.26                tidyselect_1.2.0               RSQLite_2.2.20                
#>   [6] htmlwidgets_1.6.0              grid_4.2.1                     Rtsne_0.16                     devtools_2.4.5                 munsell_0.5.0                 
#>  [11] xmlparsedata_1.0.5             codetools_0.2-18               ica_1.0-3                      future_1.30.0                  miniUI_0.1.1.1                
#>  [16] withr_2.5.0                    spatstat.random_3.0-1          colorspace_2.0-3               progressr_0.12.0               knitr_1.41                    
#>  [21] rstudioapi_0.14                Seurat_4.3.0                   ROCR_1.0-11                    tensor_1.5                     listenv_0.9.0                 
#>  [26] GenomeInfoDbData_1.2.8         polyclip_1.10-4                bit64_4.0.5                    rhdf5_2.40.0                   rprojroot_2.0.3               
#>  [31] parallelly_1.33.0              vctrs_0.5.1                    generics_0.1.3                 xfun_0.36                      R6_2.5.1                      
#>  [36] bitops_1.0-7                   rhdf5filters_1.8.0             spatstat.utils_3.0-1           cachem_1.0.6                   DelayedArray_0.22.0           
#>  [41] assertthat_0.2.1               promises_1.2.0.1               scales_1.2.1                   gtable_0.3.1                   globals_0.16.2                
#>  [46] processx_3.8.0                 goftest_1.2-3                  rlang_1.0.6                    codegrip_0.0.0.9000            splines_4.2.1                 
#>  [51] lazyeval_0.2.2                 spatstat.geom_3.0-3            yaml_2.3.6                     reshape2_1.4.4                 abind_1.4-5                   
#>  [56] httpuv_1.6.7                   tools_4.2.1                    usethis_2.1.6                  ggplot2_3.4.0                  ellipsis_0.3.2                
#>  [61] RColorBrewer_1.1-3             sessioninfo_1.2.2              ggridges_0.5.4                 Rcpp_1.0.9                     plyr_1.8.8                    
#>  [66] zlibbioc_1.42.0                purrr_1.0.0                    RCurl_1.98-1.9                 ps_1.7.2                       prettyunits_1.1.1             
#>  [71] deldir_1.0-6                   pbapply_1.6-0                  cowplot_1.1.1                  urlchecker_1.0.1               zoo_1.8-11                    
#>  [76] SeuratObject_4.1.3             ggrepel_0.9.2                  cluster_2.1.3                  fs_1.5.2                       magrittr_2.0.3                
#>  [81] data.table_1.14.6              scattermore_0.8                lmtest_0.9-40                  RANN_2.6.1                     fitdistrplus_1.1-8            
#>  [86] pkgload_1.3.2                  patchwork_1.1.2                mime_0.12                      evaluate_0.19                  xtable_1.8-4                  
#>  [91] gridExtra_2.3                  compiler_4.2.1                 tibble_3.1.8                   KernSmooth_2.23-20             crayon_1.5.2                  
#>  [96] htmltools_0.5.4                later_1.3.0                    tidyr_1.2.1                    DBI_1.1.3                      MASS_7.3-57                   
#> [101] rappdirs_0.3.3                 Matrix_1.5-3                   brio_1.1.3                     cli_3.5.0                      parallel_4.2.1                
#> [106] igraph_1.3.5                   pkgconfig_2.0.3                ttservice_0.2.2                sp_1.5-1                       plotly_4.10.1                 
#> [111] spatstat.sparse_3.0-0          xml2_1.3.3                     roxygen2_7.2.3                 XVector_0.36.0                 stringr_1.5.0                 
#> [116] callr_3.7.3                    digest_0.6.31                  sctransform_0.3.5              RcppAnnoy_0.0.20               spatstat.data_3.0-0           
#> [121] rmarkdown_2.19                 leiden_0.4.3                   uwot_0.1.14                    curl_4.3.3                     shiny_1.7.4                   
#> [126] lifecycle_1.0.3                nlme_3.1-157                   jsonlite_1.8.4                 Rhdf5lib_1.18.2                desc_1.4.2                    
#> [131] viridisLite_0.4.1              fansi_1.0.3                    pillar_1.8.1                   lattice_0.20-45                fastmap_1.1.0                 
#> [136] httr_1.4.4                     pkgbuild_1.4.0                 survival_3.3-1                 glue_1.6.2                     remotes_2.4.2                 
#> [141] png_0.1-8                      bit_4.0.5                      stringi_1.7.8                  profvis_0.3.7                  HDF5Array_1.24.2              
#> [146] blob_1.2.3                     tidySingleCellExperiment_1.6.3 memoise_2.0.1                  irlba_2.3.5.1                  future.apply_1.10.0
```
