<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Gets a Metacell from curated metadata — get_metacell • cellNexus</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gets a Metacell from curated metadata — get_metacell"><meta name="description" content="Given a data frame of Curated Atlas metadata obtained from get_metadata(),
returns a SingleCellExperiment::SingleCellExperiment object
corresponding to the samples in that data frame"><meta property="og:description" content="Given a data frame of Curated Atlas metadata obtained from get_metadata(),
returns a SingleCellExperiment::SingleCellExperiment object
corresponding to the samples in that data frame"><meta property="og:image" content="https://mangiolalaboratory.github.io/cellNexus/logo.png"></head><body><p>









  pkgdown/extra.html

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellNexus</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">home</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://j-andrews7-cellnexus.share.connect.posit.cloud/">Query builder</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/gene-expression-explore.html">gene expression explore</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MangiolaLaboratory/cellNexus/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Gets a Metacell from curated metadata</h1>
      <small class="dont-index">Source: <a href="https://github.com/MangiolaLaboratory/cellNexus/blob/master/R/counts.R" class="external-link"><code>R/counts.R</code></a></small>
      <div class="d-none name"><code>get_metacell.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given a data frame of Curated Atlas metadata obtained from <code><a href="get_metadata.html">get_metadata()</a></code>,
returns a <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment::SingleCellExperiment</a></code> object
corresponding to the samples in that data frame</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_metacell</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  assays <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  <span class="va">cell_aggregation</span>,</span>
<span>  cache_directory <span class="op">=</span> <span class="fu"><a href="get_default_cache_dir.html">get_default_cache_dir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  repository <span class="op">=</span> <span class="va">COUNTS_URL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p><a href="https://www.biorxiv.org/content/10.1101/2023.06.08.542671v3" class="external-link">Mangiola et al.,2023</a></p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame containing, at minimum, <code>sample_id</code>, <code>file_id_cellNexus_single_cell</code>,
<code>atlas_id</code> and a metacell column (e.g <code>metacell_2</code>) columns, which correspond to sample ID
file subdivision for internal use, atlas name in format (e.g cellxgene/06-02-2025)
for internal use, and metacell column to be queried.
They can be obtained from the <code><a href="get_metadata.html">get_metadata()</a></code> function.</p></dd>


<dt id="arg-assays">assays<a class="anchor" aria-label="anchor" href="#arg-assays"></a></dt>
<dd><p>A character vector of metacell counts. Default to "counts".</p></dd>


<dt id="arg-cell-aggregation">cell_aggregation<a class="anchor" aria-label="anchor" href="#arg-cell-aggregation"></a></dt>
<dd><p>A character vector representing the level of metacell aggregation.
It indicates a group of cells that can be divided into the number of metacells.
Each metacell comprises a minimum of ten single cells by default.</p></dd>


<dt id="arg-cache-directory">cache_directory<a class="anchor" aria-label="anchor" href="#arg-cache-directory"></a></dt>
<dd><p>An optional character vector of length one. If
provided, it should indicate a local file path where any remotely accessed
files should be copied.</p></dd>


<dt id="arg-repository">repository<a class="anchor" aria-label="anchor" href="#arg-repository"></a></dt>
<dd><p>A character vector of length one. If provided, it should be
an HTTP URL pointing to the location where the single cell data is stored.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>An optional character vector of features (ie genes) to return
the counts for. By default counts for all features will be returned.
When provided, the returned object will contain exactly the requested
features (row order preserved), and any experiments/samples that do not
contain all requested features are dropped. This preserves the full set
of requested features at the cost of potentially fewer samples.
A warning is emitted when samples are dropped.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>SingleCellExperiment</code> object.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Mangiola, S., M. Milton, N. Ranathunga, C. S. N. Li-Wai-Suen,
A. Odainic, E. Yang, W. Hutchison et al. "A multi-organ map of the human
immune system across age, sex and ethnicity." bioRxiv (2023): 2023-06.
doi:10.1101/2023.06.08.542671.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Use the lightweight sample database URL (for fast checks during development only)</span></span></span>
<span class="r-in"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="fu">cellNexus</span><span class="fu">::</span><span class="va"><a href="SAMPLE_DATABASE_URL.html">SAMPLE_DATABASE_URL</a></span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Downloading 1 file, totalling 0 GB</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-metadata/sample_metadata.1.3.0.parquet to /home/runner/.cache/R/cellNexus/sample_metadata.1.3.0.parquet</span>
<span class="r-in"><span><span class="va">metacell</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">|&gt;</span> <span class="fu">get_metacell</span><span class="op">(</span>cell_aggregation <span class="op">=</span> <span class="st">"metacell_2"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Synchronising files</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Downloading 1 file, totalling 0 GB</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Downloading https://object-store.rc.nectar.org.au/v1/AUTH_06d6e008e3e642da99d806ba3ea629c5/cellNexus-anndata/cellxgene/21-08-2025/metacell_2/counts/03319e4f54220f534de2c4e42e607126___1.h5ad to /home/runner/.cache/R/cellNexus/cellxgene/21-08-2025/metacell_2/counts/03319e4f54220f534de2c4e42e607126___1.h5ad</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Reading files.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Compiling Experiment.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stefano Mangiola, Mengyuan Shen, Michael Milton, Silicon Valley Foundation CZF2019-002443, NIH NHGRI 5U24HG004059-18, Victoria Cancer Agency ECRF21036, NHMRC 1116955.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  
</body></html>

