<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Metadata Explore • cellNexus</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Metadata Explore">
</head>
<body>
<p>








  pkgdown/extra.html

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellNexus</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">home</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://j-andrews7-cellnexus.share.connect.posit.cloud/">Query builder</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/gene-expression-explore.html">gene expression explore</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MangiolaLaboratory/cellNexus/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Metadata Explore</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MangiolaLaboratory/cellNexus/blob/master/vignettes/metadata-explore.Rmd" class="external-link"><code>vignettes/metadata-explore.Rmd</code></a></small>
      <div class="d-none name"><code>metadata-explore.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="why-this-page-exists">Why this page exists<a class="anchor" aria-label="anchor" href="#why-this-page-exists"></a>
</h2>
<p>This page is a standalone metadata guide for <code>cellNexus</code>
and documents the key fields used in downstream analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MangiolaLaboratory/cellNexus" class="external-link">cellNexus</a></span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_metadata.html">get_metadata</a></span><span class="op">(</span>cloud_metadata <span class="op">=</span> <span class="va">SAMPLE_DATABASE_URL</span><span class="op">)</span></span>
<span><span class="va">metadata</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-processing-context">Data-processing context<a class="anchor" aria-label="anchor" href="#data-processing-context"></a>
</h2>
<p><code>cellNexus</code> metadata are harmonised to support
cross-dataset analysis:</p>
<ul>
<li>Common ontology-backed labels are retained where possible.</li>
<li>Additional curated columns support quality control and robust
grouping.</li>
<li>Expression retrieval APIs use metadata filters to provide
analysis-ready objects.</li>
</ul>
</div>
<div class="section level2">
<h2 id="metadata-dictionary">Metadata dictionary<a class="anchor" aria-label="anchor" href="#metadata-dictionary"></a>
</h2>
<div class="section level3">
<h3 id="dataset-level-fields">Dataset-level fields<a class="anchor" aria-label="anchor" href="#dataset-level-fields"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dataset_id</code></td>
<td>Primary dataset identifier used in the atlas.</td>
</tr>
<tr class="even">
<td><code>dataset_version_id</code></td>
<td>Versioned identifier for a dataset release.</td>
</tr>
<tr class="odd">
<td><code>collection_id</code></td>
<td>Parent collection identifier grouping related datasets.</td>
</tr>
<tr class="even">
<td><code>cell_count</code></td>
<td>Number of cells associated with the dataset.</td>
</tr>
<tr class="odd">
<td><code>filetype</code></td>
<td>Storage format of source expression files.</td>
</tr>
<tr class="even">
<td><code>is_primary_data</code></td>
<td>Flags whether data are marked as primary observations.</td>
</tr>
<tr class="odd">
<td><code>mean_genes_per_cell</code></td>
<td>Mean detected genes per cell reported for the dataset.</td>
</tr>
<tr class="even">
<td><code>published_at</code></td>
<td>Publication timestamp for dataset release.</td>
</tr>
<tr class="odd">
<td><code>revised_at</code></td>
<td>Last revision timestamp for the dataset metadata.</td>
</tr>
<tr class="even">
<td><code>schema_version</code></td>
<td>CELLxGENE schema version associated with this record.</td>
</tr>
<tr class="odd">
<td><code>tombstone</code></td>
<td>Indicates whether a dataset has been retired/deprecated
upstream.</td>
</tr>
<tr class="even">
<td><code>x_normalization</code></td>
<td>Original normalization/scale annotation from source metadata.</td>
</tr>
<tr class="odd">
<td><code>explorer_url</code></td>
<td>URL to browse the dataset in a public explorer.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="sample-level-fields">Sample-level fields<a class="anchor" aria-label="anchor" href="#sample-level-fields"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>sample_id</code></td>
<td>Harmonised sample identifier used in <code>cellNexus</code>.</td>
</tr>
<tr class="even">
<td><code>sample_</code></td>
<td>Source sample label from upstream metadata.</td>
</tr>
<tr class="odd">
<td><code>donor_id</code></td>
<td>Donor or participant identifier linked to the sample.</td>
</tr>
<tr class="even">
<td><code>age_days</code></td>
<td>Donor age represented in days (harmonised).</td>
</tr>
<tr class="odd">
<td><code>sex</code></td>
<td>Recorded biological sex label.</td>
</tr>
<tr class="even">
<td><code>sex_ontology_term_id</code></td>
<td>Ontology identifier for <code>sex</code>.</td>
</tr>
<tr class="odd">
<td><code>development_stage</code></td>
<td>Developmental stage label.</td>
</tr>
<tr class="even">
<td><code>development_stage_ontology_term_id</code></td>
<td>Ontology identifier for <code>development_stage</code>.</td>
</tr>
<tr class="odd">
<td><code>self_reported_ethnicity</code></td>
<td>Self-reported ancestry/ethnicity label where available.</td>
</tr>
<tr class="even">
<td><code>self_reported_ethnicity_ontology_term_id</code></td>
<td>Ontology identifier for self-reported ethnicity.</td>
</tr>
<tr class="odd">
<td><code>organism</code></td>
<td>Organism name for the sample (for example, human).</td>
</tr>
<tr class="even">
<td><code>organism_ontology_term_id</code></td>
<td>Ontology identifier for <code>organism</code>.</td>
</tr>
<tr class="odd">
<td><code>assay</code></td>
<td>Assay/platform label used for sequencing.</td>
</tr>
<tr class="even">
<td><code>assay_ontology_term_id</code></td>
<td>Ontology identifier for <code>assay</code>.</td>
</tr>
<tr class="odd">
<td><code>tissue</code></td>
<td>Tissue label used for biological grouping.</td>
</tr>
<tr class="even">
<td><code>tissue_ontology_term_id</code></td>
<td>Ontology identifier for <code>tissue</code>.</td>
</tr>
<tr class="odd">
<td><code>tissue_type</code></td>
<td>Tissue class (for example, tissue vs organoid context).</td>
</tr>
<tr class="even">
<td><code>tissue_groups</code></td>
<td>Curated high-level grouping of tissues.</td>
</tr>
<tr class="odd">
<td><code>disease</code></td>
<td>Disease/condition annotation.</td>
</tr>
<tr class="even">
<td><code>disease_ontology_term_id</code></td>
<td>Ontology identifier for <code>disease</code>.</td>
</tr>
<tr class="odd">
<td><code>sample_placeholder</code></td>
<td>Placeholder marker from upstream records where needed.</td>
</tr>
<tr class="even">
<td><code>experiment___</code></td>
<td>Upstream experiment grouping variable.</td>
</tr>
<tr class="odd">
<td><code>is_immune</code></td>
<td>Curated flag indicating immune-cell context.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="cell-level-fields">Cell-level fields<a class="anchor" aria-label="anchor" href="#cell-level-fields"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>cell_id</code></td>
<td>Unique cell barcode/identifier.</td>
</tr>
<tr class="even">
<td><code>observation_joinid</code></td>
<td>Join key used to link metadata and expression records.</td>
</tr>
<tr class="odd">
<td><code>cell_type</code></td>
<td>Source/curated cell-type label.</td>
</tr>
<tr class="even">
<td><code>cell_type_ontology_term_id</code></td>
<td>Ontology identifier for <code>cell_type</code>.</td>
</tr>
<tr class="odd">
<td><code>cell_type_unified_ensemble</code></td>
<td>
<code>cellNexus</code> consensus immune-cell identity from
multi-method annotation.</td>
</tr>
<tr class="even">
<td><code>cell_annotation_azimuth_l2</code></td>
<td>Azimuth level-2 annotation used in harmonisation.</td>
</tr>
<tr class="odd">
<td><code>cell_annotation_blueprint_singler</code></td>
<td>SingleR annotation using Blueprint reference.</td>
</tr>
<tr class="even">
<td><code>cell_annotation_blueprint_monaco</code></td>
<td>SingleR annotation using Monaco reference.</td>
</tr>
<tr class="odd">
<td><code>empty_droplet</code></td>
<td>Quality-control flag for probable empty droplets.</td>
</tr>
<tr class="even">
<td><code>alive</code></td>
<td>Quality-control flag for viable/non-damaged cells.</td>
</tr>
<tr class="odd">
<td><code>scDblFinder.class</code></td>
<td>Doublet classification label from <code>scDblFinder</code>.</td>
</tr>
<tr class="even">
<td><code>nCount_RNA</code></td>
<td>Total RNA counts detected in a cell (sample-aware).</td>
</tr>
<tr class="odd">
<td><code>nFeature_expressed_in_sample</code></td>
<td>Number of expressed features detected in a cell.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="cellnexus-infrastructure-fields">
<code>cellNexus</code> infrastructure fields<a class="anchor" aria-label="anchor" href="#cellnexus-infrastructure-fields"></a>
</h3>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th>Column</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>sample_heuristic</code></td>
<td>Internal sample subdivision/grouping helper.</td>
</tr>
<tr class="even">
<td><code>file_id_cellNexus_single_cell</code></td>
<td>Internal file identifier for single-cell layers.</td>
</tr>
<tr class="odd">
<td><code>file_id_cellNexus_pseudobulk</code></td>
<td>Internal file identifier for pseudobulk layers.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="practical-exploration">Practical exploration<a class="anchor" aria-label="anchor" href="#practical-exploration"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Which columns are available?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many datasets per tissue?</span></span>
<span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">dataset_id</span>, <span class="va">tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">tissue</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Typical quality-control filtering</span></span>
<span><span class="va">metadata_qc</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">empty_droplet</span> <span class="op">==</span> <span class="cn">FALSE</span>,</span>
<span>    <span class="va">alive</span> <span class="op">==</span> <span class="cn">TRUE</span>,</span>
<span>    <span class="va">scDblFinder.class</span> <span class="op">!=</span> <span class="st">"doublet"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Stefano Mangiola, Mengyuan Shen, Michael Milton, Silicon Valley Foundation CZF2019-002443, NIH NHGRI 5U24HG004059-18, Victoria Cancer Agency ECRF21036, NHMRC 1116955.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  
</body>
</html>
